File: src/sequence_basics.py
"""Basic helpers for synthetic DNA sequence examples."""

import pandas as pd

DNA_COMPLEMENT = str.maketrans("ACGTacgt", "TGCAtgca")


def clean_sequence(sequence: str) -> str:
    """Return an uppercase sequence without spaces or line breaks."""
    return "".join(str(sequence).split()).upper()


def gc_content(sequence: str) -> float:
    """Calculate GC content as a percent."""
    seq = clean_sequence(sequence)
    if not seq:
        raise ValueError("sequence must not be empty.")
    gc = seq.count("G") + seq.count("C")
    return gc / len(seq) * 100


def reverse_complement(sequence: str) -> str:
    """Return the reverse complement of a DNA sequence."""
    seq = clean_sequence(sequence)
    invalid = set(seq) - set("ACGT")
    if invalid:
        raise ValueError(f"Invalid DNA bases: {sorted(invalid)}")
    return seq.translate(DNA_COMPLEMENT)[::-1]


def count_motif(sequence: str, motif: str) -> int:
    """Count non-overlapping motif occurrences in a DNA sequence."""
    seq = clean_sequence(sequence)
    motif = clean_sequence(motif)
    if not motif:
        raise ValueError("motif must not be empty.")
    return seq.count(motif)


def summarize_sequences(df: pd.DataFrame, sequence_col: str = "dna_sequence") -> pd.DataFrame:
    """Add length and GC-content summaries to a sequence table."""
    if sequence_col not in df.columns:
        raise ValueError(f"Missing required columns: {[sequence_col]}")
    result = df.copy()
    result["sequence_length"] = result[sequence_col].map(lambda seq: len(clean_sequence(seq)))
    result["gc_content_percent"] = result[sequence_col].map(gc_content)
    return result

File: R/sequence_basics.R
library(tidyverse)

# Basic helpers for synthetic DNA sequence examples.

clean_sequence <- function(sequence) {
  toupper(gsub("\\s+", "", as.character(sequence)))
}

gc_content <- function(sequence) {
  seq <- clean_sequence(sequence)
  if (nchar(seq) == 0) {
    stop("sequence must not be empty")
  }
  chars <- strsplit(seq, "")[[1]]
  (sum(chars %in% c("G", "C")) / length(chars)) * 100
}

count_motif <- function(sequence, motif) {
  seq <- clean_sequence(sequence)
  motif <- clean_sequence(motif)
  if (nchar(motif) == 0) {
    stop("motif must not be empty")
  }
  stringr::str_count(seq, fixed(motif))
}

reverse_complement <- function(sequence) {
  seq <- clean_sequence(sequence)
  invalid <- setdiff(strsplit(seq, "")[[1]], c("A", "C", "G", "T"))
  if (length(invalid) > 0) {
    stop(paste("Invalid DNA bases:", paste(unique(invalid), collapse = ", ")))
  }
  comp <- chartr("ACGT", "TGCA", seq)
  paste(rev(strsplit(comp, "")[[1]]), collapse = "")
}

summarize_sequences <- function(df, sequence_col = "dna_sequence") {
  if (!sequence_col %in% names(df)) {
    stop(paste("Missing required columns:", sequence_col))
  }
  df %>%
    mutate(
      sequence_length = nchar(map_chr(.data[[sequence_col]], clean_sequence)),
      gc_content_percent = map_dbl(.data[[sequence_col]], gc_content)
    )
}
